snowflake/snowflake_test.go

```diff
@@ -0,0 +1,230 @@
+	assert.Error(err)
+}
+
+func Test_SetNodeIDProvider_Ok(t *testing.T) {
+	g := &Generator{
+		Settings: Settings{
+			Epoch:            time.Now(),
+			SequenceResolver: AtomicResolver,
+		},
+	}
+
+	assert := assert.New(t)
+	err := g.setNodeIDProvider(customProvider)
+	assert.NoError(err)
+
+	generated, err := g.NextID()
+	assert.NoError(err)
+	decomposed := g.Decompose(generated)
+	assert.Equal(nodeID, decomposed.NodeID)
+
+	err = g.setNodeIDProvider(customProviderPlus)
+	assert.NoError(err)
+	generated, err = g.NextID()
+	assert.NoError(err)
+	decomposed = g.Decompose(generated)
+	assert.Equal(nodeID, decomposed.NodeID)
+}
+
+func newID(t *testing.T) uint64 {
+	assert := assert.New(t)
+
+	id, err := NextID()
+	if err != nil {
+		t.Fatal("ID not generated")
+	}
+	assert.NotZero(id)
+	return id
+}
+
+func currentTime() time.Time {
+	return common.CurrentUTCTime()
+}
+
+func Test_NextIDFor10Sec(t *testing.T) {
+	assert := assert.New(t)
+	sequenceBits := 8
+	var numID uint32
+	var lastID uint64
+	var maxSequence uint16
+	nodeID, err := host.Lower16BitPrivateIP()
+	assert.NoError(err)
+
+	initial := currentTime()
+	current := initial
+	for current.Sub(initial) < 10*time.Second {
+		id := newID(t)
+		parts := Decompose(id)
+		numID++
+
+		if id <= lastID {
+			t.Fatal("duplicated ID")
+		}
+		lastID = id
+
+		current = currentTime()
+
+		actualMSB := parts.MSB
+		if actualMSB != 0 {
+			t.Errorf("unexpected msb: %d", actualMSB)
+		}
+
+		actualTime := parts.Timestamp
+		overtime := actualTime.Sub(current)
+		if overtime > 0 {
+			t.Errorf("unexpected overtime: %d", overtime)
+		}
+
+		actualSequence := parts.Sequence
+		if maxSequence < actualSequence {
+			maxSequence = actualSequence
+		}
+
+		actualNodeID := parts.NodeID
+		if actualNodeID != nodeID {
+			t.Errorf("unexpected nodeID: %d", actualNodeID)
+		}
+	}
+
+	if maxSequence > 1<<sequenceBits-1 {
+		t.Errorf("unexpected max sequence: %d", maxSequence)
+	}
+}
+
+func Test_NewIDInParallel(t *testing.T) {
+	numCPU := runtime.NumCPU()
+	runtime.GOMAXPROCS(numCPU)
+
+	const numID = 10000
+	numGenerator := 1
+	if numCPU/2 > numGenerator {
+		numGenerator = numCPU / 2
+	}
+
+	generate := func(wg *sync.WaitGroup, partition []uint64) {
+		wg.Add(1)
+		defer wg.Done()
+		start := time.Now()
+		for i := 0; i < numID; i++ {
+			partition[i] = NewID()
+
+		}
+		elapsed := time.Since(start)
+		fmt.Printf("%v IDs generaged, tooks %v s\n", numID, elapsed)
+	}
+	wg := sync.WaitGroup{}
+	start := time.Now()
+	ids := make(map[int][]uint64, numGenerator)
+	for i := 0; i < numGenerator; i++ {
+		ids[i] = make([]uint64, numID)
+		go generate(&wg, ids[i])
+	}
+	wg.Wait()
+	elapsed := time.Since(start)
+	fmt.Printf("%v IDs generaged, tooks %v, speed: %v IDs/sec\n", numID*numGenerator,
+		elapsed, float64(numID*numGenerator)/elapsed.Seconds())
+
+	set := make(map[uint64]int)
+	for idx, partition := range ids {
+		for i := 0; i < numID; i++ {
+			id := partition[i]
+			if idx, ok := set[id]; ok {
```

Comment:
Recreate the variable name `idx` make me confused in the error below (there are 2 `idx` in the error format string)

Reply:
fixed